04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,77 @@
+const Product = require("../models/product");
+
+const getAllProductsStatic = async (req, res) => {
+  const products = await Product.find({ price: { $gt: 30 } })
+    .sort("price")
+    .select("name price");
+  // .limit(10)
+  // .skip(1);
+
+  // throw new Error('testing async errors')
+  res.status(200).json({ products, nbHits: products.length });
+};
+
+const getAllProducts = async (req, res) => {
+  const { featured, company, name, sort, fields, numericFilters } = req.query;
+  const queryObject = {};
+
+  if (featured) {
+    queryObject.featured = featured === "true" ? true : false;
+  }
+  if (company) {
+    queryObject.company = company;
+  }
+  if (name) {
+    queryObject.name = { $regex: name, $options: "i" };
+  }
+  if (numericFilters) {
+    const operatorMap = {
+      ">": "$gt",
+      ">=": "$gte",
+      "=": "$eq",
+      "<": "$lt",
+      "<=": "$lte",
+    };
+    const regEx = /\b(<|>|<=|=|>=)\b/g;
+    let filters = numericFilters.replace(
+      regEx,
+      (match) => `-${operatorMap[match]}-`
+    );
+    const options = ["price", "rating"];
+    filters = filters.split(",").forEach((item) => {
+      const [field, operator, value] = item.split("-");
+      if (options.includes(field)) {
+        queryObject[field] = { [operator]: Number(value) };
+      }
+    });
+  }
+
+  // console.log(queryObject)
+  let result = Product.find(queryObject);
+  // sort
+  if (sort) {
+    const sortList = sort.split(",").join(" ");
+    result = result.sort(sortList);
+  } else {
+    result = result.sort("createdAt");
+  }
+  if (fields) {
+    const fieldsList = fields.split(",").join(" ");
+    result = result.select(fieldsList);
+  }
+  const page = Number(req.query.page) || 1;
+  const limit = Number(req.query.limit) || 10;
+  const skip = (page - 1) * limit;
+
+  result = result.skip(skip).limit(limit);
```

Comment:
The reassignment of the `result` variable isn't needed on lines  54, 60, or 66.
So this could just be:
`result.skip(skip).limit(limit);`
without changing the functionality.

This is due to Mongoose's API using lone query method calls like a builder, where it stores each query call and only executes them once an `await` is used or `.then()` is used. This is covered a little bit in the documentation:
[https://mongoosejs.com/docs/queries.html#executing](https://mongoosejs.com/docs/queries.html#executing)

and you can see an example from my mentor code on it as well:
[https://github.com/joroctd/node-express-course/blob/mw8/04-store-api/starter/controllers/products.js](https://github.com/joroctd/node-express-course/blob/mw8/04-store-api/starter/controllers/products.js)